{"nbformat":4,"nbformat_minor":0,"metadata":{"colab":{"provenance":[],"authorship_tag":"ABX9TyO/Rn1gYE0bGKfxZQXCRCrG"},"kernelspec":{"name":"python3","display_name":"Python 3"},"language_info":{"name":"python"}},"cells":[{"cell_type":"code","execution_count":null,"metadata":{"id":"1ZaBwH1REu2s"},"outputs":[],"source":["# generating sorted lists of NCBI numbers and sequences from gene.json for protein and nucleotide\n","import pandas as pd\n","import json\n","import re\n","\n","with open('gene.json') as f:\n","    raw_data = json.load(f)\n","try:\n","      # normalizing data (without \"success\")\n","      data_normalized = pd.json_normalize(raw_data['success'])\n","except Exception as e:\n","      print(f\"An error occurred during normalization: {e}\")\n","\n","df_ncbi = data_normalized[[\"gene_ncbi\",\"sequence\"]]\n","display(df_ncbi)\n","#6754 rows\n","\n","# 1. filter out without NCBI number and cleaning GI number format\n","pattern = r'^GI:[A-Za-z0-9]+$'\n","mask_invalid_format = ~df_ncbi['gene_ncbi'].str.match(pattern, na=False)\n","df_valid = df_ncbi[~mask_invalid_format].copy()\n","display(df_valid)\n","#6752 rows\n","df_valid['gene_ncbi'] = df_valid['gene_ncbi'].str.replace('^GI:', '', regex=True)\n","df_valid['sequence'] = df_valid['sequence'].str.upper()\n","display(df_valid)\n","\n","# 2. without sequence, split into p and n\n","\n","mask_na_or_empty = df_valid['sequence'].isna() | (df_valid['sequence'].str.strip() == '')\n","df_valid_no_seq = df_valid[mask_na_or_empty].copy()\n","display(df_valid_no_seq)\n","#434\n","\n","# Split based on the last letter\n","df_valid_no_seq['last_char'] = df_valid_no_seq['gene_ncbi'].str[-1]\n","\n","df_p_no_seq = df_valid_no_seq[df_valid_no_seq['last_char'] == 'p'].drop(columns='last_char')\n","df_p_no_seq = df_p_no_seq['gene_ncbi']\n","df_p_no_seq = df_p_no_seq.map(lambda x: ''.join(filter(str.isdigit, str(x))))\n","display(df_p_no_seq)\n","#366 rows\n","# to file\n","df_p_no_seq.to_csv(\"df_p_no_seq.csv\", index=False)\n","\n","df_no_letter_no_seq = df_valid_no_seq[~df_valid_no_seq['last_char'].isin(['p', 'n'])].drop(columns='last_char')\n","df_no_letter_no_seq = df_no_letter_no_seq['gene_ncbi']\n","display(df_no_letter_no_seq)\n","#68 rows\n","# to file\n","df_no_letter_no_seq.to_csv(\"df_no_letter_no_seq.csv\", index=False)\n","# no_letter changed to n in the local file name\n","\n","# 3. with sequence to fasta file, split into p and n\n","\n","mask_na_or_empty = df_valid['sequence'].isna() | (df_valid['sequence'].str.strip() == '')\n","df_valid_seq = df_valid[~mask_na_or_empty].copy()\n","display(df_valid_seq)\n","#6318 rows\n","\n","df_valid_seq['last_char'] = df_valid_seq['gene_ncbi'].str[-1]\n","\n","df_p = df_valid_seq[df_valid_seq['last_char'] == 'p'].drop(columns='last_char')\n","df_p['gene_ncbi'] = df_p['gene_ncbi'].map(lambda x: ''.join(filter(str.isdigit, str(x))))\n","display(df_p)\n","#2342 rows\n","# to file\n","df_p.to_csv(\"df_p.csv\", index=False)\n","\n","df_no_letter = df_valid_seq[~df_valid_seq['last_char'].isin(['p', 'n'])].drop(columns='last_char')\n","display(df_no_letter)\n","#3976 rows\n","# to file\n","df_no_letter.to_csv(\"df_no_letter.csv\", index=False)\n","# no_letter changed to n in the local file name"]},{"cell_type":"code","source":["# analysis of information from E-Direct, success of obtaining accession number, duplicates between protein, nucleotide, gene, remaining records for searching\n","\n","import pandas as pd\n","import re\n","\n","# full_list protein vs full_list nucleotide vs full_list genes\n","# repeat for full_list and no_seq_list\n","\n","df_p = pd.read_csv('output_GI_to_accesion_protein_emailversion.tsv', sep='\\t')\n","df_n = pd.read_csv('output_GI_to_accesion_nucleotide_emailversion.tsv', sep='\\t')\n","df_g = pd.read_csv('output_GI_to_accesion_gene_emailversion.tsv', sep='\\t')\n","#pd.DataFrame(columns=[\"GI\", \"AccessionVersion\"])\n","\n","#display(df_p)\n","#display(df_n)\n","#display(df_g)\n","\n","\n","# what part of protein was mapped\n","missing_count = df_p['AccessionVersion'].isna().sum() + (df_p['AccessionVersion'].astype(str).str.strip() == '').sum()\n","mask_na_or_empty = df_p['AccessionVersion'].isna() | (df_p['AccessionVersion'].str.strip() == 'nan')\n","df_p_filtered = df_p[~mask_na_or_empty].copy()\n","df_p_filtered.to_csv(\"df_p_filtered.csv\", index=False)\n","print(\"Protein DB\")\n","print(f\"Number of empty cells: {missing_count}\")\n","print(f\"Total number of rows: {len(df_p)}\")\n","print(f\"Number of retrieved accession numbers: {len(df_p_filtered)}\")\n","\n","# what part of nucleotide was mapped\n","missing_count = df_n['AccessionVersion'].isna().sum() + (df_n['AccessionVersion'].astype(str).str.strip() == '').sum()\n","mask_na_or_empty = df_n['AccessionVersion'].isna() | (df_n['AccessionVersion'].str.strip() == 'nan')\n","df_n_filtered = df_n[~mask_na_or_empty].copy()\n","df_n_filtered.to_csv(\"df_n_filtered.csv\", index=False)\n","print(\"Nucleotide DB\")\n","print(f\"Number of empty cells: {missing_count}\")\n","print(f\"Total number of rows: {len(df_n)}\")\n","print(f\"Number of retrieved accession numbers: {len(df_n_filtered)}\")\n","\n","# what part of gene was mapped\n","missing_count = df_g['AccessionVersion'].isna().sum() + (df_g['AccessionVersion'].astype(str).str.strip() == '').sum()\n","mask_na_or_empty = df_g['AccessionVersion'].isna() | (df_g['AccessionVersion'].astype(str).str.strip() == 'nan')\n","df_g_filtered = df_g[~mask_na_or_empty].copy()\n","df_g_filtered.to_csv(\"df_g_filtered.csv\", index=False)\n","print(\"Gene DB\")\n","print(f\"Number of empty cells: {missing_count}\")\n","print(f\"Total number of rows: {len(df_g)}\")\n","print(f\"Number of retrieved accession numbers: {len(df_g_filtered)}\")\n","\n","\n","\n","# Compare columns between protein, nucleotide, genes accession mapping results = are there duplicate mappings\n","setp = set(df_p_filtered['GI'].dropna().astype(str).str.strip())\n","setn = set(df_n_filtered['GI'].dropna().astype(str).str.strip())\n","setg = set(df_g_filtered['GI'].dropna().astype(str).str.strip())\n","\n","\n","common_pn = setp & setn\n","common_pg = setp & setg\n","common_ng = setn & setg\n","\n","pn_only_p = setp - setn\n","pn_only_n = setn - setp\n","pg_only_p = setp - setg\n","pg_only_g = setg - setp\n","ng_only_n = setn - setg\n","ng_only_g = setg - setn\n","\n","\n","print(f\"Common values pn: {len(common_pn)}\")\n","print(f\"Common values pg: {len(common_pg)}\")\n","print(f\"Common values ng: {len(common_ng)}\")\n","\n","print(f\"PN only in p: {len(pn_only_p)}\")\n","print(f\"PN only in n: {len(pn_only_n)}\")\n","print(f\"PG only in p: {len(pg_only_p)}\")\n","print(f\"PG only in g: {len(pg_only_g)}\")\n","print(f\"NG only in n: {len(ng_only_n)}\")\n","print(f\"NG only in g: {len(ng_only_g)}\")"],"metadata":{"id":"yRVTcyBSFqyW"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["import pandas as pd\n","import re\n","\n","# lists of p and n AccessionVersion to download fasta from NCBI (filter out those for which I have sequences) + missing record\n","\n","# df_p_filtered and df_n_filtered from the previous block are two-column dfs with GI and AccessionVersion\n","df_p_filtered = pd.read_csv('df_p_filtered.csv')\n","df_n_filtered = pd.read_csv('df_n_filtered.csv')\n","df_p_clean_list = pd.read_csv('df_p.csv') # columns gene_ncbi,sequence\n","df_p_no_seq_clean_list = pd.read_csv('df_p_no_seq.csv') # column gene_ncbi\n","df_n_clean_list = pd.read_csv('df_n.csv') # columns gene_ncbi,sequence\n","df_n_no_seq_clean_list = pd.read_csv('df_n_no_seq.csv') # column gene_ncbi\n","\n","# unifying column names\n","\n","df_p_clean_list = df_p_clean_list.rename(columns={'gene_ncbi': 'GI'})\n","df_p_no_seq_clean_list = df_p_no_seq_clean_list.rename(columns={'gene_ncbi': 'GI'})\n","df_n_clean_list = df_n_clean_list.rename(columns={'gene_ncbi': 'GI'})\n","df_n_no_seq_clean_list = df_n_no_seq_clean_list.rename(columns={'gene_ncbi': 'GI'})\n","\n","# adding AccessionNumber = merge to a 3-column df\n","\n","df_p_clean_list = pd.merge(df_p_clean_list, df_p_filtered, on='GI', how='left')\n","df_p_no_seq_clean_list = pd.merge(df_p_no_seq_clean_list, df_p_filtered, on='GI', how='left')\n","df_n_clean_list = pd.merge(df_n_clean_list, df_n_filtered, on='GI', how='left')\n","df_n_no_seq_clean_list = pd.merge(df_n_no_seq_clean_list, df_n_filtered, on='GI', how='left')\n","\n","display(df_n_filtered)\n","display(df_p_clean_list)\n","display(df_p_no_seq_clean_list)\n","display(df_n_clean_list)\n","display(df_n_no_seq_clean_list)\n","\n","# extracting those NaNs that were supposed to be p but are n (df_n_clean_list)\n","\n","mask_na_or_empty = df_n_clean_list['AccessionVersion'].isna() | (df_n_clean_list['AccessionVersion'].astype(str).str.strip() == 'nan')\n","df_n_nan = df_n_clean_list[mask_na_or_empty].copy()\n","#display(df_n_nan)\n","df_n_nan = df_n_nan.drop(columns=['AccessionVersion'])\n","df_n_nan = pd.merge(df_n_nan, df_p_filtered, on='GI', how='left')\n","df_n_nan['NCBI_DB'] = df_n_nan['AccessionVersion'].notna().map({True: 'P', False: ''})\n","display(df_n_nan)\n","\n","df_n_clean_list = df_n_clean_list[~mask_na_or_empty]\n","display(df_n_clean_list)\n","df_n_clean_list['NCBI_DB'] = df_n_clean_list['AccessionVersion'].notna().map({True: 'N', False: ''})\n","display(df_n_clean_list)\n","\n","df_n_clean_list_ncbi_db = pd.concat([df_n_clean_list, df_n_nan], ignore_index=True)\n","display(df_n_clean_list_ncbi_db)\n","\n","# clean lists of AccessionVersion to download KO\n","\n","df_KO_p_all = pd.concat([df_p_clean_list['AccessionVersion'], df_p_no_seq_clean_list['AccessionVersion'], df_n_nan['AccessionVersion']], ignore_index=True)\n","display(df_KO_p_all)\n","\n","df_KO_n_all = pd.concat([df_n_clean_list['AccessionVersion'], df_n_no_seq_clean_list['AccessionVersion']], ignore_index=True)\n","display(df_KO_n_all)\n","\n","df_KO_p_all.to_csv(\"df_KO_p_all.csv\", index=False)\n","df_KO_n_all.to_csv(\"df_KO_n_all.csv\", index=False)\n","\n","# df for generating fasta directly from file\n","# df_p_clean_list\n","df_p_clean_list['sequence'] = df_p_clean_list['sequence'].str.replace(r'\\s+', '', regex=True)\n","df_p_clean_list.to_csv(\"df_p_clean_list.csv\", index=False)\n","# df_n_clean_list\n","df_n_clean_list_ncbi_db['sequence'] = df_n_clean_list_ncbi_db['sequence'].str.replace(r'\\s+', '', regex=True)\n","df_n_clean_list_ncbi_db.to_csv(\"df_n_clean_list_ncbi_db.csv\", index=False)"],"metadata":{"id":"NOsmK5AWF0P9"},"execution_count":null,"outputs":[]},{"cell_type":"code","source":["# script for generating FASTA from sequences in BioSurfDB (+AccessionNumber)\n","df_p_clean_list = pd.read_csv('df_p_clean_list.csv')\n","df_n_clean_list_ncbi_db = pd.read_csv('df_n_clean_list_ncbi_db.csv')\n","\n","# generating fasta for df_p_clean_list\n","with open('output_p.fasta', 'w') as fasta_file:\n","    for _, row in df_p_clean_list.iterrows():\n","        id_part = f\"{str(row['GI']).strip()},{str(row['AccessionVersion']).strip()}\"\n","        sequence = str(row['sequence']).strip()\n","\n","        if id_part and sequence:\n","            fasta_file.write(f\">{id_part}\\n{sequence}\\n\")\n","\n","# generating fasta for df_n_clean_list_ncbi_db\n","with open('output_n.fasta', 'w') as fasta_file:\n","    for _, row in df_n_clean_list_ncbi_db.iterrows():\n","        id_part = f\"{str(row['GI']).strip()},{str(row['AccessionVersion']).strip()},{str(row['NCBI_DB']).strip()}\"\n","        sequence = str(row['sequence']).strip()\n","\n","        if id_part and sequence:\n","            fasta_file.write(f\">{id_part}\\n{sequence}\\n\")"],"metadata":{"id":"gynjHJZBF1VN"},"execution_count":null,"outputs":[]}]}