{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": []
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "NiGaFY-o_S1Y"
      },
      "outputs": [],
      "source": [
        "import matplotlib.pyplot as plt\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import json"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#change display settings for whole dfs\n",
        "pd.set_option('display.max_rows', None)       # Display all rows\n",
        "pd.set_option('display.max_columns', None)    # Display all columns\n",
        "pd.set_option('display.width', None)          # Disable line wrapping\n",
        "pd.set_option('display.max_colwidth', None)   # Show full column width"
      ],
      "metadata": {
        "id": "w4PuuMrPCPE4"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#reading json files downloaded from biosurfdb\n",
        "#list of file \"name\" -> df\n",
        "name_list = [\n",
        "    \"organism\",\n",
        "    \"gene\",\n",
        "    \"protein\",\n",
        "    \"pathway\",\n",
        "    \"surfactant\",\n",
        "    \"metagenome\",\n",
        "    \"biosurfactant\",\n",
        "    \"organism_pathway\"\n",
        "]\n",
        "\n",
        "#loading data into df inside a dict for all files\n",
        "df_dict = {}\n",
        "for name in name_list:\n",
        "\n",
        "  # Load json\n",
        "  with open(f'{name}.json') as f:\n",
        "    raw_data = json.load(f)\n",
        "\n",
        "  try:\n",
        "      # normalizing data (without success)\n",
        "      data_normalized = pd.json_normalize(raw_data['success'])\n",
        "  except Exception as e:\n",
        "      print(f\"An error occurred during normalization: {e}\")\n",
        "  #Save to dict with proper name\n",
        "  df_dict[f\"{name}_df\"] = data_normalized\n",
        "\n",
        "  #Show df\n",
        "  display(df_dict[f\"{name}_df\"].head())"
      ],
      "metadata": {
        "id": "X2OYp7OAQbr-"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#merge dfs into df_combined\n",
        "\n",
        "#biosurfactant+surfactant\n",
        "df_dict[\"biosurfactant_surfactant_df_combined\"] = pd.merge(df_dict[\"biosurfactant_df\"], df_dict[\"surfactant_df\"], on='surfactant_id', how='inner')\n",
        "\n",
        "#organism_pathway+additional columns\n",
        "  #0.change column names to be common\n",
        "  #gene_df:  gene_ncbi -> gene_GI\n",
        "  #pathway_df: name -> pathway\n",
        "  #protein_df: protein_ncbi -> protein_GI\n",
        "  #organism_df: binomial_name -> organism\n",
        "df_dict['gene_df'].rename(columns={'gene_ncbi': 'Gene_GI'}, inplace=True)\n",
        "df_dict['pathway_df'].rename(columns={'name': 'Pathway'}, inplace=True)\n",
        "df_dict['protein_df'].rename(columns={'protein_ncbi': 'Protein_GI'}, inplace=True)\n",
        "df_dict['organism_df'].rename(columns={'binomial_name': 'Organism'}, inplace=True)\n",
        "\n",
        "df_dict[\"organism_pathway_df_combined\"] = pd.merge(df_dict[\"organism_pathway_df\"], df_dict[\"organism_df\"], on='Organism', how='left')\n",
        "df_dict[\"organism_pathway_df_combined\"] = pd.merge(df_dict[\"organism_pathway_df\"], df_dict[\"gene_df\"], on='Gene_GI', how='left')\n",
        "df_dict[\"organism_pathway_df_combined\"] = pd.merge(df_dict[\"organism_pathway_df\"], df_dict[\"pathway_df\"], on='Pathway', how='left')\n",
        "df_dict[\"organism_pathway_df_combined\"] = pd.merge(df_dict[\"organism_pathway_df\"], df_dict[\"protein_df\"], on='Protein_GI', how='left')\n",
        "\n",
        "#metagenome+organism_pathway by reference -> filter\n",
        "df_dict[\"metagenome_df\"].rename(columns={'reference': 'Reference'}, inplace=True)\n",
        "#filter None in Reference in metagenome_df\n",
        "df_dict[\"metagenome_df\"] = df_dict[\"metagenome_df\"][df_dict[\"metagenome_df\"][\"Reference\"].notna()]\n",
        "df_dict[\"metagenome_df\"] = df_dict[\"metagenome_df\"][['environment', 'Reference']]\n",
        "display(df_dict[\"metagenome_df\"])\n",
        "df_dict[\"metagenome_organism_pathway_df_combined\"] = pd.merge(df_dict[\"metagenome_df\"], df_dict[\"organism_pathway_df\"], on='Reference', how='left')\n",
        "\n",
        "#pathway+biosurfactant by reference\n",
        "\n",
        "df_dict[\"pathway_biosurfactant_df_combined\"] = pd.merge(df_dict[\"pathway_df\"], df_dict[\"biosurfactant_df\"], on='reference', how='inner')"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 424
        },
        "collapsed": true,
        "id": "gZJ2f8zjkjVb",
        "outputId": "dc786a12-779a-4630-8450-b979ac856139"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "                                environment      Reference\n",
              "0    0-10 cm of soil (excluding crust) from  PMID:19601957\n",
              "1                           150 km offshore  PMID:23671678\n",
              "2             1959 Kilauea volcanic deposit               \n",
              "3             2,6-dichlorophenol enrichment               \n",
              "4      2-3 km off coast sea water, Valencia               \n",
              "..                                      ...            ...\n",
              "728                        Yaizu Hot Spring               \n",
              "729                                  yogurt  PMID:15292125\n",
              "730              Zootermopsis angusticollis  PMID:21541298\n",
              "731                      Zoothamnium niveum  PMID:19120466\n",
              "732                          Zostera marina  PMID:11760947\n",
              "\n",
              "[733 rows x 2 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-5d64b82d-498c-47a3-a1c2-b71172ee76e6\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>environment</th>\n",
              "      <th>Reference</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>0-10 cm of soil (excluding crust) from</td>\n",
              "      <td>PMID:19601957</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>150 km offshore</td>\n",
              "      <td>PMID:23671678</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>1959 Kilauea volcanic deposit</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>2,6-dichlorophenol enrichment</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>2-3 km off coast sea water, Valencia</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>728</th>\n",
              "      <td>Yaizu Hot Spring</td>\n",
              "      <td></td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>729</th>\n",
              "      <td>yogurt</td>\n",
              "      <td>PMID:15292125</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>730</th>\n",
              "      <td>Zootermopsis angusticollis</td>\n",
              "      <td>PMID:21541298</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>731</th>\n",
              "      <td>Zoothamnium niveum</td>\n",
              "      <td>PMID:19120466</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>732</th>\n",
              "      <td>Zostera marina</td>\n",
              "      <td>PMID:11760947</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>733 rows × 2 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-5d64b82d-498c-47a3-a1c2-b71172ee76e6')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-5d64b82d-498c-47a3-a1c2-b71172ee76e6 button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-5d64b82d-498c-47a3-a1c2-b71172ee76e6');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-c5ee1836-bb8c-4a81-8ccd-e28b0072122e\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-c5ee1836-bb8c-4a81-8ccd-e28b0072122e')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-c5ee1836-bb8c-4a81-8ccd-e28b0072122e button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "summary": "{\n  \"name\": \"df_dict[\\\"pathway_biosurfactant_df_combined\\\"] = pd\",\n  \"rows\": 733,\n  \"fields\": [\n    {\n      \"column\": \"environment\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 733,\n        \"samples\": [\n          \"tissue\",\n          \"activated sludge from oil refinery\",\n          \"Grapevine\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Reference\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 224,\n        \"samples\": [\n          \"PMID:24500796\",\n          \"PMID:23861734\",\n          \"PMID:23458364\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#Display combined dfs\n",
        "display(df_dict['organism_pathway_df_combined'])\n",
        "display(df_dict['metagenome_organism_pathway_df_combined'])\n",
        "display(df_dict['pathway_biosurfactant_df_combined'])"
      ],
      "metadata": {
        "collapsed": true,
        "id": "UVc8cMu0upFQ",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 652
        },
        "outputId": "8c85e5ba-4d9b-4026-d85c-9ba5c093dbd5"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "display_data",
          "data": {
            "text/plain": [
              "                              Organism                 Strain  \\\n",
              "0             Psychrobacter glacincola                   kzT3   \n",
              "1                   Ostreococcus tauri               OTTH0595   \n",
              "2         Omnitrophica bacterium OLB16                  OLB16   \n",
              "3                    Bacillus subtilis                    ZK8   \n",
              "4                    Bacillus subtilis                    ZK0   \n",
              "...                                ...                    ...   \n",
              "11937  Geobacillus thermodenitrificans                R-32622   \n",
              "11938  Geobacillus thermodenitrificans                R-32623   \n",
              "11939  Geobacillus thermodenitrificans                R-32624   \n",
              "11940  Geobacillus thermodenitrificans                R-32506   \n",
              "11941         Geobacillus kaustophilus  type strain: LMG 9819   \n",
              "\n",
              "                                                    Gene         Gene_GI  \\\n",
              "0                                                   almA    GI:679035587   \n",
              "1      Kynurenine 3-monooxygenase and related        ...    GI:308809476   \n",
              "2                                           unknown gene  GI:1000276437p   \n",
              "3                                           unknown gene  GI:1001619011p   \n",
              "4                                           unknown gene  GI:1001619014p   \n",
              "...                                                  ...             ...   \n",
              "11937                                               nosZ    GI:576080337   \n",
              "11938                                               nosZ    GI:576080339   \n",
              "11939                                               nosZ    GI:576080341   \n",
              "11940                                               nosZ    GI:576080345   \n",
              "11941                                               nosZ    GI:576080319   \n",
              "\n",
              "                                      Protein               Protein_GI  \\\n",
              "0                flavin-binding monooxygenase               CDD:224983   \n",
              "1      Kynurenine 3-monooxygenase and related           GeneID:9835880   \n",
              "2                   Muconate cycloisomerase 1            GI:1000276437   \n",
              "3                                        ItuC            GI:1001619011   \n",
              "4                                        ItuA            GI:1001619014   \n",
              "...                                       ...                      ...   \n",
              "11937                 nitrous oxide reductase  UniProtKB/TrEMBL:W0UVS1   \n",
              "11938                 nitrous oxide reductase  UniProtKB/TrEMBL:W0UVS1   \n",
              "11939                 nitrous oxide reductase  UniProtKB/TrEMBL:W0UVS1   \n",
              "11940                 nitrous oxide reductase  UniProtKB/TrEMBL:W0UVS1   \n",
              "11941                 nitrous oxide reductase  UniProtKB/TrEMBL:W0UVT9   \n",
              "\n",
              "                           Pathway       Family      Reference  \\\n",
              "0           Fatty acid degradation  Degradation                  \n",
              "1      Generic Degradation Pathway  Degradation  PMID:16868079   \n",
              "2             Benzoate degradation  Degradation  PMID:27029554   \n",
              "3            Iturin A Biosynthesis  Surfactants                  \n",
              "4            Iturin A Biosynthesis  Surfactants                  \n",
              "...                            ...          ...            ...   \n",
              "11937          Nitrogen metabolism     Nitrogen  PMID:24784780   \n",
              "11938          Nitrogen metabolism     Nitrogen  PMID:24784780   \n",
              "11939          Nitrogen metabolism     Nitrogen  PMID:24784780   \n",
              "11940          Nitrogen metabolism     Nitrogen  PMID:24784780   \n",
              "11941          Nitrogen metabolism     Nitrogen  PMID:24784780   \n",
              "\n",
              "                                 protein_name   protein_alias  \\\n",
              "0                flavin-binding monooxygenase      AIM52630.1   \n",
              "1      Kynurenine 3-monooxygenase and related  XP_003082048.1   \n",
              "2                   Muconate cycloisomerase 1      KXK38553.1   \n",
              "3                                        ItuC      AMM04302.1   \n",
              "4                                        ItuA      AMM04304.1   \n",
              "...                                       ...             ...   \n",
              "11937                 nitrous oxide reductase      CDG32597.1   \n",
              "11938                 nitrous oxide reductase      CDG32597.1   \n",
              "11939                 nitrous oxide reductase      CDG32597.1   \n",
              "11940                 nitrous oxide reductase      CDG32597.1   \n",
              "11941                 nitrous oxide reductase      CDG32591.1   \n",
              "\n",
              "                                                sequence  \n",
              "0      GGTWDLFDYPGIRSDSDALTFGYRFRPWLDHKMLAGGADIKRYIAD...  \n",
              "1      MSDDDDGDANAGATTTTTTTTTKKLGKKERMKRRAKELVAQGVDED...  \n",
              "2      mkitgvhliplvyptrgrfkffesptgrtgraavlvkvatdeglag...  \n",
              "3      msefkqqelfwgrmfhtedrpgafpsfqvsdstvkrditsgsdcih...  \n",
              "4      mytsqfqtlvdvirersnisdrgirfiesdknetvvsyrqlfeeaq...  \n",
              "...                                                  ...  \n",
              "11937  FATDVANSRAAVMDLKTFTVKDIIEVPNTSGPHCAAFVTENTEYLF...  \n",
              "11938  FATDVANSRAAVMDLKTFTVKDIIEVPNTSGPHCAAFVTENTEYLF...  \n",
              "11939  FATDVANSRAAVMDLKTFTVKDIIEVPNTSGPHCAAFVTENTEYLF...  \n",
              "11940  FATDVANSRAAVMDLKTFTVKDIIEVPNTSGPHCAAFVTENTEYLF...  \n",
              "11941  FATDVANSRAAVMDLKTFTVKDMIEVPNTSGPHCAAFVTENTEYLF...  \n",
              "\n",
              "[11942 rows x 12 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-9a8b6b3a-dd15-4583-b3d2-883cd87998af\" class=\"colab-df-container\">\n",
              "    <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Organism</th>\n",
              "      <th>Strain</th>\n",
              "      <th>Gene</th>\n",
              "      <th>Gene_GI</th>\n",
              "      <th>Protein</th>\n",
              "      <th>Protein_GI</th>\n",
              "      <th>Pathway</th>\n",
              "      <th>Family</th>\n",
              "      <th>Reference</th>\n",
              "      <th>protein_name</th>\n",
              "      <th>protein_alias</th>\n",
              "      <th>sequence</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Psychrobacter glacincola</td>\n",
              "      <td>kzT3</td>\n",
              "      <td>almA</td>\n",
              "      <td>GI:679035587</td>\n",
              "      <td>flavin-binding monooxygenase</td>\n",
              "      <td>CDD:224983</td>\n",
              "      <td>Fatty acid degradation</td>\n",
              "      <td>Degradation</td>\n",
              "      <td></td>\n",
              "      <td>flavin-binding monooxygenase</td>\n",
              "      <td>AIM52630.1</td>\n",
              "      <td>GGTWDLFDYPGIRSDSDALTFGYRFRPWLDHKMLAGGADIKRYIAD...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Ostreococcus tauri</td>\n",
              "      <td>OTTH0595</td>\n",
              "      <td>Kynurenine 3-monooxygenase and related        ...</td>\n",
              "      <td>GI:308809476</td>\n",
              "      <td>Kynurenine 3-monooxygenase and related</td>\n",
              "      <td>GeneID:9835880</td>\n",
              "      <td>Generic Degradation Pathway</td>\n",
              "      <td>Degradation</td>\n",
              "      <td>PMID:16868079</td>\n",
              "      <td>Kynurenine 3-monooxygenase and related</td>\n",
              "      <td>XP_003082048.1</td>\n",
              "      <td>MSDDDDGDANAGATTTTTTTTTKKLGKKERMKRRAKELVAQGVDED...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>Omnitrophica bacterium OLB16</td>\n",
              "      <td>OLB16</td>\n",
              "      <td>unknown gene</td>\n",
              "      <td>GI:1000276437p</td>\n",
              "      <td>Muconate cycloisomerase 1</td>\n",
              "      <td>GI:1000276437</td>\n",
              "      <td>Benzoate degradation</td>\n",
              "      <td>Degradation</td>\n",
              "      <td>PMID:27029554</td>\n",
              "      <td>Muconate cycloisomerase 1</td>\n",
              "      <td>KXK38553.1</td>\n",
              "      <td>mkitgvhliplvyptrgrfkffesptgrtgraavlvkvatdeglag...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>Bacillus subtilis</td>\n",
              "      <td>ZK8</td>\n",
              "      <td>unknown gene</td>\n",
              "      <td>GI:1001619011p</td>\n",
              "      <td>ItuC</td>\n",
              "      <td>GI:1001619011</td>\n",
              "      <td>Iturin A Biosynthesis</td>\n",
              "      <td>Surfactants</td>\n",
              "      <td></td>\n",
              "      <td>ItuC</td>\n",
              "      <td>AMM04302.1</td>\n",
              "      <td>msefkqqelfwgrmfhtedrpgafpsfqvsdstvkrditsgsdcih...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>Bacillus subtilis</td>\n",
              "      <td>ZK0</td>\n",
              "      <td>unknown gene</td>\n",
              "      <td>GI:1001619014p</td>\n",
              "      <td>ItuA</td>\n",
              "      <td>GI:1001619014</td>\n",
              "      <td>Iturin A Biosynthesis</td>\n",
              "      <td>Surfactants</td>\n",
              "      <td></td>\n",
              "      <td>ItuA</td>\n",
              "      <td>AMM04304.1</td>\n",
              "      <td>mytsqfqtlvdvirersnisdrgirfiesdknetvvsyrqlfeeaq...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11937</th>\n",
              "      <td>Geobacillus thermodenitrificans</td>\n",
              "      <td>R-32622</td>\n",
              "      <td>nosZ</td>\n",
              "      <td>GI:576080337</td>\n",
              "      <td>nitrous oxide reductase</td>\n",
              "      <td>UniProtKB/TrEMBL:W0UVS1</td>\n",
              "      <td>Nitrogen metabolism</td>\n",
              "      <td>Nitrogen</td>\n",
              "      <td>PMID:24784780</td>\n",
              "      <td>nitrous oxide reductase</td>\n",
              "      <td>CDG32597.1</td>\n",
              "      <td>FATDVANSRAAVMDLKTFTVKDIIEVPNTSGPHCAAFVTENTEYLF...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11938</th>\n",
              "      <td>Geobacillus thermodenitrificans</td>\n",
              "      <td>R-32623</td>\n",
              "      <td>nosZ</td>\n",
              "      <td>GI:576080339</td>\n",
              "      <td>nitrous oxide reductase</td>\n",
              "      <td>UniProtKB/TrEMBL:W0UVS1</td>\n",
              "      <td>Nitrogen metabolism</td>\n",
              "      <td>Nitrogen</td>\n",
              "      <td>PMID:24784780</td>\n",
              "      <td>nitrous oxide reductase</td>\n",
              "      <td>CDG32597.1</td>\n",
              "      <td>FATDVANSRAAVMDLKTFTVKDIIEVPNTSGPHCAAFVTENTEYLF...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11939</th>\n",
              "      <td>Geobacillus thermodenitrificans</td>\n",
              "      <td>R-32624</td>\n",
              "      <td>nosZ</td>\n",
              "      <td>GI:576080341</td>\n",
              "      <td>nitrous oxide reductase</td>\n",
              "      <td>UniProtKB/TrEMBL:W0UVS1</td>\n",
              "      <td>Nitrogen metabolism</td>\n",
              "      <td>Nitrogen</td>\n",
              "      <td>PMID:24784780</td>\n",
              "      <td>nitrous oxide reductase</td>\n",
              "      <td>CDG32597.1</td>\n",
              "      <td>FATDVANSRAAVMDLKTFTVKDIIEVPNTSGPHCAAFVTENTEYLF...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11940</th>\n",
              "      <td>Geobacillus thermodenitrificans</td>\n",
              "      <td>R-32506</td>\n",
              "      <td>nosZ</td>\n",
              "      <td>GI:576080345</td>\n",
              "      <td>nitrous oxide reductase</td>\n",
              "      <td>UniProtKB/TrEMBL:W0UVS1</td>\n",
              "      <td>Nitrogen metabolism</td>\n",
              "      <td>Nitrogen</td>\n",
              "      <td>PMID:24784780</td>\n",
              "      <td>nitrous oxide reductase</td>\n",
              "      <td>CDG32597.1</td>\n",
              "      <td>FATDVANSRAAVMDLKTFTVKDIIEVPNTSGPHCAAFVTENTEYLF...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>11941</th>\n",
              "      <td>Geobacillus kaustophilus</td>\n",
              "      <td>type strain: LMG 9819</td>\n",
              "      <td>nosZ</td>\n",
              "      <td>GI:576080319</td>\n",
              "      <td>nitrous oxide reductase</td>\n",
              "      <td>UniProtKB/TrEMBL:W0UVT9</td>\n",
              "      <td>Nitrogen metabolism</td>\n",
              "      <td>Nitrogen</td>\n",
              "      <td>PMID:24784780</td>\n",
              "      <td>nitrous oxide reductase</td>\n",
              "      <td>CDG32591.1</td>\n",
              "      <td>FATDVANSRAAVMDLKTFTVKDMIEVPNTSGPHCAAFVTENTEYLF...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>11942 rows × 12 columns</p>\n",
              "</div>\n",
              "    <div class=\"colab-df-buttons\">\n",
              "\n",
              "  <div class=\"colab-df-container\">\n",
              "    <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-9a8b6b3a-dd15-4583-b3d2-883cd87998af')\"\n",
              "            title=\"Convert this dataframe to an interactive table.\"\n",
              "            style=\"display:none;\">\n",
              "\n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\" viewBox=\"0 -960 960 960\">\n",
              "    <path d=\"M120-120v-720h720v720H120Zm60-500h600v-160H180v160Zm220 220h160v-160H400v160Zm0 220h160v-160H400v160ZM180-400h160v-160H180v160Zm440 0h160v-160H620v160ZM180-180h160v-160H180v160Zm440 0h160v-160H620v160Z\"/>\n",
              "  </svg>\n",
              "    </button>\n",
              "\n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    .colab-df-buttons div {\n",
              "      margin-bottom: 4px;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "    <script>\n",
              "      const buttonEl =\n",
              "        document.querySelector('#df-9a8b6b3a-dd15-4583-b3d2-883cd87998af button.colab-df-convert');\n",
              "      buttonEl.style.display =\n",
              "        google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "      async function convertToInteractive(key) {\n",
              "        const element = document.querySelector('#df-9a8b6b3a-dd15-4583-b3d2-883cd87998af');\n",
              "        const dataTable =\n",
              "          await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                    [key], {});\n",
              "        if (!dataTable) return;\n",
              "\n",
              "        const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "          '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "          + ' to learn more about interactive tables.';\n",
              "        element.innerHTML = '';\n",
              "        dataTable['output_type'] = 'display_data';\n",
              "        await google.colab.output.renderOutput(dataTable, element);\n",
              "        const docLink = document.createElement('div');\n",
              "        docLink.innerHTML = docLinkHtml;\n",
              "        element.appendChild(docLink);\n",
              "      }\n",
              "    </script>\n",
              "  </div>\n",
              "\n",
              "\n",
              "    <div id=\"df-f5b89a96-0622-4c06-a6c1-50430e984462\">\n",
              "      <button class=\"colab-df-quickchart\" onclick=\"quickchart('df-f5b89a96-0622-4c06-a6c1-50430e984462')\"\n",
              "                title=\"Suggest charts\"\n",
              "                style=\"display:none;\">\n",
              "\n",
              "<svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "     width=\"24px\">\n",
              "    <g>\n",
              "        <path d=\"M19 3H5c-1.1 0-2 .9-2 2v14c0 1.1.9 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2zM9 17H7v-7h2v7zm4 0h-2V7h2v10zm4 0h-2v-4h2v4z\"/>\n",
              "    </g>\n",
              "</svg>\n",
              "      </button>\n",
              "\n",
              "<style>\n",
              "  .colab-df-quickchart {\n",
              "      --bg-color: #E8F0FE;\n",
              "      --fill-color: #1967D2;\n",
              "      --hover-bg-color: #E2EBFA;\n",
              "      --hover-fill-color: #174EA6;\n",
              "      --disabled-fill-color: #AAA;\n",
              "      --disabled-bg-color: #DDD;\n",
              "  }\n",
              "\n",
              "  [theme=dark] .colab-df-quickchart {\n",
              "      --bg-color: #3B4455;\n",
              "      --fill-color: #D2E3FC;\n",
              "      --hover-bg-color: #434B5C;\n",
              "      --hover-fill-color: #FFFFFF;\n",
              "      --disabled-bg-color: #3B4455;\n",
              "      --disabled-fill-color: #666;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart {\n",
              "    background-color: var(--bg-color);\n",
              "    border: none;\n",
              "    border-radius: 50%;\n",
              "    cursor: pointer;\n",
              "    display: none;\n",
              "    fill: var(--fill-color);\n",
              "    height: 32px;\n",
              "    padding: 0;\n",
              "    width: 32px;\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart:hover {\n",
              "    background-color: var(--hover-bg-color);\n",
              "    box-shadow: 0 1px 2px rgba(60, 64, 67, 0.3), 0 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "    fill: var(--button-hover-fill-color);\n",
              "  }\n",
              "\n",
              "  .colab-df-quickchart-complete:disabled,\n",
              "  .colab-df-quickchart-complete:disabled:hover {\n",
              "    background-color: var(--disabled-bg-color);\n",
              "    fill: var(--disabled-fill-color);\n",
              "    box-shadow: none;\n",
              "  }\n",
              "\n",
              "  .colab-df-spinner {\n",
              "    border: 2px solid var(--fill-color);\n",
              "    border-color: transparent;\n",
              "    border-bottom-color: var(--fill-color);\n",
              "    animation:\n",
              "      spin 1s steps(1) infinite;\n",
              "  }\n",
              "\n",
              "  @keyframes spin {\n",
              "    0% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "      border-left-color: var(--fill-color);\n",
              "    }\n",
              "    20% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    30% {\n",
              "      border-color: transparent;\n",
              "      border-left-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    40% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-top-color: var(--fill-color);\n",
              "    }\n",
              "    60% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "    }\n",
              "    80% {\n",
              "      border-color: transparent;\n",
              "      border-right-color: var(--fill-color);\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "    90% {\n",
              "      border-color: transparent;\n",
              "      border-bottom-color: var(--fill-color);\n",
              "    }\n",
              "  }\n",
              "</style>\n",
              "\n",
              "      <script>\n",
              "        async function quickchart(key) {\n",
              "          const quickchartButtonEl =\n",
              "            document.querySelector('#' + key + ' button');\n",
              "          quickchartButtonEl.disabled = true;  // To prevent multiple clicks.\n",
              "          quickchartButtonEl.classList.add('colab-df-spinner');\n",
              "          try {\n",
              "            const charts = await google.colab.kernel.invokeFunction(\n",
              "                'suggestCharts', [key], {});\n",
              "          } catch (error) {\n",
              "            console.error('Error during call to suggestCharts:', error);\n",
              "          }\n",
              "          quickchartButtonEl.classList.remove('colab-df-spinner');\n",
              "          quickchartButtonEl.classList.add('colab-df-quickchart-complete');\n",
              "        }\n",
              "        (() => {\n",
              "          let quickchartButtonEl =\n",
              "            document.querySelector('#df-f5b89a96-0622-4c06-a6c1-50430e984462 button');\n",
              "          quickchartButtonEl.style.display =\n",
              "            google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "        })();\n",
              "      </script>\n",
              "    </div>\n",
              "\n",
              "    </div>\n",
              "  </div>\n"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "dataframe",
              "summary": "{\n  \"name\": \"display(df_dict['organism_pathway_df_combined'])\",\n  \"rows\": 11942,\n  \"fields\": [\n    {\n      \"column\": \"Organism\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 2098,\n        \"samples\": [\n          \"Bradyrhizobium japonicum USDA 6\",\n          \"Bacillus thuringiensis serovar navarrensis\",\n          \"Methanobrevibacter curvatus\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Strain\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 3336,\n        \"samples\": [\n          \"F1\",\n          \"DSM 13819\",\n          \"DSM 9756\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Gene\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 987,\n        \"samples\": [\n          \"hapE\",\n          \"tfdB\",\n          \"pheA2\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Gene_GI\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 6368,\n        \"samples\": [\n          \"GI:1890334\",\n          \"GI:107015898\",\n          \"GI:338222424\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Protein\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 1519,\n        \"samples\": [\n          \"TrbA\",\n          \"CbzT\",\n          \"dibenzothiophene sulfone monooxygenase\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Protein_GI\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 6271,\n        \"samples\": [\n          \"GI:389886441\",\n          \"GI:595674\",\n          \"GI:209863934\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Pathway\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 51,\n        \"samples\": [\n          \"Lichenysin Biosynthesis\",\n          \"Surfactin Biosynthesis\",\n          \"Alasan Biosynthesis\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Family\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 5,\n        \"samples\": [\n          \"Surfactants\",\n          \"Sulfur\",\n          \"Nitrogen\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"Reference\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 1194,\n        \"samples\": [\n          \"PMID:12501294\",\n          \"PMID:15109718\",\n          \"PMID:23376198\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"protein_name\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 1519,\n        \"samples\": [\n          \"TrbA\",\n          \"CbzT\",\n          \"dibenzothiophene sulfone monooxygenase\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"protein_alias\",\n      \"properties\": {\n        \"dtype\": \"string\",\n        \"num_unique_values\": 6262,\n        \"samples\": [\n          \"BAJ11520.1\",\n          \"WP_007784268.1\",\n          \"AKL96892.1\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    },\n    {\n      \"column\": \"sequence\",\n      \"properties\": {\n        \"dtype\": \"category\",\n        \"num_unique_values\": 5923,\n        \"samples\": [\n          \"KGLVPDGYVEGWKKTFEEDFSPRRGAELVARAWTDPDFRQLLLTDGTAAVAQYGYLGPQGEYIVAVEDTPTLKNVIVCSLCSCTAWPILGLPPTWYKSFEYRARVVREPRKVLSEMGTEIASDVEIRVYDTTAETRYMVLPQRPAAPKAGARNNCRKSSPKDCS\",\n          \"MAFDPNNLMQDRITDIGPPKYDKFFPPVIKENFGKWKYHEILEPGVLVHVSETGAEVYSVRIGSARLISIDYIREVCDIADKYCGGYLRWTTRNNIEFLVSDKTKLQPLLADLRSRGNWFPIGGTGNSITNIVHTQGWAHCHTPAIDASGVVKAVMDELFGYFGSHKLPAQIRIALACCLNMCGAVHCSDIAILGVHRKPPFVENERVQNVCEIPLVVAACPTAAIKPKKVGELKSIEINNERC\",\n          \"migrcthvvdcretsgmgkggslaqrgtfaecgtdvcavamspgrrhitkpvceitfglreanvltstmvlnagagvphdapaaggtlfgltdrevqqmdnfkllvihlggvanhiiykarlilrnidkhciiicespvdhedfakigvktskvmpkdediktegtiddivtgvirgetisqekldeiirkvklalgda\"\n        ],\n        \"semantic_type\": \"\",\n        \"description\": \"\"\n      }\n    }\n  ]\n}"
            }
          },
          "metadata": {}
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "#plot 1 (table) - Biosurfactants classification by classes\n",
        "\n",
        "biosurfactant_surfactant_df_combined_cleaned = df_dict['biosurfactant_surfactant_df_combined']\n",
        "map = {\n",
        "    'Polymeric Surfactants': 'Polymeric biosurfactants',\n",
        "    'Polymeric surfactants': 'Polymeric biosurfactants',\n",
        "    'Glycolipid': 'Glycolipids',\n",
        "    'Lipopeptide': 'Lipopeptides',\n",
        "    ' lipopeptide': 'Lipopeptides',\n",
        "    'Biosurfactant': 'Other biosurfactants',\n",
        "    'Biosurfactants': 'Other biosurfactants'\n",
        "}\n",
        "biosurfactant_surfactant_df_combined_cleaned['class_name'] = biosurfactant_surfactant_df_combined_cleaned['class_name'].replace(map)\n",
        "biosurf_by_class = biosurfactant_surfactant_df_combined_cleaned.groupby('class_name')['surfactant_name'].nunique()\n",
        "display(biosurf_by_class)\n",
        "biosurf_by_class_plot = plt.bar(biosurf_by_class.index, biosurf_by_class.values, color='orange')\n",
        "plt.xticks(rotation=90)\n",
        "plt.xlabel(\"Clase de biosurfactante\")\n",
        "plt.ylabel(\"Cantidad de biosurfactantes\")\n",
        "plt.title(\"Cantidad de biosurfactantes por clase\")\n",
        "plt.savefig(\"biosurfactants_by_class.png\", dpi=300, bbox_inches='tight')\n",
        "plt.show(biosurf_by_class_plot)"
      ],
      "metadata": {
        "collapsed": true,
        "id": "f_WD2W-MkwQ_"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#PLOT 2.1 - Distribution of biosurfactants production ability in bacterial genera (pathways)\n",
        "pathways_in_organism = df_dict['organism_pathway_df_combined'].groupby('Pathway')['Organism'].nunique()\n",
        "display(pathways_in_organism)\n",
        "#normalized version\n",
        "pathways_in_organism = pathways_in_organism[pathways_in_organism.values < 100]\n",
        "\n",
        "pathways_in_organism_plot = plt.bar(pathways_in_organism.index, pathways_in_organism.values, color='orange')\n",
        "plt.xticks(rotation=90, fontsize=6)\n",
        "plt.xlabel(\"Pathway\")\n",
        "plt.ylabel(\"Number of organisms\")\n",
        "plt.title(\"Number of organisms by pathway\")\n",
        "plt.savefig(\"organisms_by_pathway_normalized.png\", dpi=300, bbox_inches='tight')\n",
        "plt.show(pathways_in_organism_plot)"
      ],
      "metadata": {
        "id": "qN_3-KyBKkfH"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#PLOT 2.2 - Distribution of  biosurfactants production ability in metagenomes (pathways)\n",
        "\n",
        "nr_pathways_in_metagenome = df_dict['organism_pathway_df_combined'].groupby('Pathway')['Reference'].nunique()\n",
        "#normalized version\n",
        "nr_pathways_in_metagenome = nr_pathways_in_metagenome[nr_pathways_in_metagenome.values < 30]\n",
        "\n",
        "display(nr_pathways_in_metagenome)\n",
        "nr_pathways_in_metagenome_plot = plt.bar(nr_pathways_in_metagenome.index, nr_pathways_in_metagenome.values, color='orange')\n",
        "plt.xticks(rotation=90, fontsize=6)\n",
        "plt.xlabel(\"Pathway\")\n",
        "plt.ylabel(\"Number of metagenomes\")\n",
        "plt.title(\"Number of metagenomes by pathway\")\n",
        "plt.savefig(\"metagenomes_by_pathway_normalized.png\", dpi=300, bbox_inches='tight')\n",
        "plt.show(nr_pathways_in_metagenome_plot)\n",
        "\n",
        "#reverse plot\n",
        "\n",
        "nr_pathways_in_metagenome_reverse = df_dict['organism_pathway_df_combined'].groupby('Reference')['Pathway'].nunique()\n",
        "#normalized version\n",
        "nr_pathways_in_metagenome_reverse = nr_pathways_in_metagenome_reverse[nr_pathways_in_metagenome_reverse.values > 3]\n",
        "nr_pathways_in_metagenome_reverse = nr_pathways_in_metagenome_reverse[nr_pathways_in_metagenome_reverse.values < 10]\n",
        "\n",
        "display(nr_pathways_in_metagenome_reverse)\n",
        "nr_pathways_in_metagenome_reverse_plot = plt.bar(nr_pathways_in_metagenome_reverse.index, nr_pathways_in_metagenome_reverse.values, color='orange')\n",
        "plt.xticks(rotation=90, fontsize=6)\n",
        "plt.xlabel(\"Metagenome\")\n",
        "plt.ylabel(\"Number of pathways\")\n",
        "plt.title(\"Number of pathways by metagenomes\")\n",
        "plt.savefig(\"metagenomes_by_pathway_reverse_metagenomelist.png\", dpi=300, bbox_inches='tight')\n",
        "plt.show(nr_pathways_in_metagenome_reverse_plot)"
      ],
      "metadata": {
        "id": "c4c9DREVLQ2y"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#PLOT 3 - Amount of genes included in metabolical pathways\n",
        "nr_genes_in_pathway = df_dict['organism_pathway_df_combined'].groupby('Pathway')['Gene_GI'].nunique()\n",
        "#normalized version\n",
        "#nr_genes_in_pathway = nr_genes_in_pathway[nr_genes_in_pathway.values < 200]\n",
        "\n",
        "nr_genes_in_pathway_plot = plt.bar(nr_genes_in_pathway.index, nr_genes_in_pathway.values, color='orange')\n",
        "plt.xticks(rotation=90, fontsize=6)\n",
        "plt.xlabel(\"Ruta metabólica\")\n",
        "plt.ylabel(\"Cantidad de genes\")\n",
        "plt.title(\"Cantidad de genes por ruta metabólica\")\n",
        "plt.savefig(\"genes_by_pathway_normalized.png\", dpi=300, bbox_inches='tight')\n",
        "plt.show(nr_genes_in_pathway_plot)\n",
        "\n",
        "#plot 3_colored\n",
        "\n",
        "nr_genes_in_pathway = df_dict['organism_pathway_df_combined'].groupby('Pathway')['Gene_GI'].nunique()\n",
        "\n",
        "# Create map Pathway → Family\n",
        "pathway_family_map = df_dict['organism_pathway_df_combined'].drop_duplicates(subset='Pathway')[['Pathway', 'Family']].set_index('Pathway')['Family']\n",
        "# Make sure indexes match\n",
        "families = pathway_family_map.loc[nr_genes_in_pathway.index]\n",
        "\n",
        "# Assign colors to unique families\n",
        "import seaborn as sns\n",
        "unique_families = families.unique()\n",
        "palette = sns.color_palette(\"colorblind\", n_colors=len(unique_families))\n",
        "color_map = dict(zip(unique_families, palette))\n",
        "bar_colors = families.map(color_map)\n",
        "\n",
        "# Plot\n",
        "plt.figure(figsize=(10, 6))\n",
        "bars = plt.bar(nr_genes_in_pathway.index, nr_genes_in_pathway.values, color=bar_colors)\n",
        "\n",
        "plt.xticks(rotation=90, fontsize=6)\n",
        "plt.xlabel(\"Ruta metabólica\")\n",
        "plt.ylabel(\"Cantidad de genes\")\n",
        "plt.title(\"Cantidad de genes por ruta metabólica\")\n",
        "\n",
        "from matplotlib.patches import Patch\n",
        "legend_handles = [Patch(color=color_map[f], label=f) for f in unique_families]\n",
        "plt.legend(handles=legend_handles, title='Familia')\n",
        "\n",
        "plt.savefig(\"genes_by_pathway_colored.png\", dpi=300, bbox_inches='tight')\n",
        "plt.show(nr_genes_in_pathway_plot)"
      ],
      "metadata": {
        "id": "lABkNeo4Kka3"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#PLOT metagenome_by_organism\n",
        "#plot 4.X** - Distribution of biosurfactants producing bacteria in metagenomes\n",
        "\n",
        "nr_metagenomes_by_organism = df_dict['organism_pathway_df_combined'].groupby('Organism')['Reference'].nunique()\n",
        "\n",
        "nr_metagenomes_by_organism_normalized = nr_metagenomes_by_organism[nr_metagenomes_by_organism.values > 4]\n",
        "display(nr_metagenomes_by_organism_normalized)\n",
        "\n",
        "nr_metagenomes_by_organism_normalized_plot = plt.bar(nr_metagenomes_by_organism_normalized.index, nr_metagenomes_by_organism_normalized.values, color='orange')\n",
        "plt.xticks(rotation=90, fontsize=6)\n",
        "plt.xlabel(\"Organism\")\n",
        "plt.ylabel(\"Number of metagenomes\")\n",
        "plt.title(\"Number of metagenomes by organism\")\n",
        "plt.savefig(\"metagenomes_by_organism.png\", dpi=300, bbox_inches='tight')\n",
        "plt.show(nr_metagenomes_by_organism_normalized_plot)\n",
        "\n",
        "#cleaned by genera version\n",
        "\n",
        "nr_metagenomes_by_organism_cleaned = df_dict['organism_pathway_df_combined']\n",
        "\n",
        "map = {\n",
        "    'Acinetobacter baumannii': 'Acinetobacter sp.',\n",
        "    'Acinetobacter calcoaceticus': 'Acinetobacter sp.',\n",
        "    'Alcaligenes faecalis': 'Alcaligenes sp.',\n",
        "    'Bacillus subtilis': 'Bacillus sp.',\n",
        "    'Bacillus licheniformis': 'Bacillus sp.',\n",
        "    'Burkholderia cepacia': 'Burkholderia sp.',\n",
        "    'Comamonas testosteroni': 'Comamonas sp.',\n",
        "    'Methylobacterium extorquens': 'Methylobacterium sp.',\n",
        "    'Pseudomonas aeruginosa': 'Pseudomonas sp.',\n",
        "    'Pseudomonas fluorescens': 'Pseudomonas sp.',\n",
        "    'Pseudomonas pseudoalcaligenes': 'Pseudomonas sp.',\n",
        "    'Pseudomonas putida': 'Pseudomonas sp.',\n",
        "    'Pseudomonas stutzeri': 'Pseudomonas sp.',\n",
        "    'Ralstonia eutropha': 'Ralstonia sp.',\n",
        "    'Ralstonia pickettii': 'Ralstonia sp.',\n",
        "    'Rhodococcus erythropolis': 'Rhodococcus sp.',\n",
        "    'Rhodococcus jostii': 'Rhodococcus sp.',\n",
        "    'Rhodococcus opacus': 'Rhodococcus sp.',\n",
        "    'Rhodococcus rhodochrous': 'Rhodococcus sp.',\n",
        "    'Sphingomonas paucimobilis': 'Sphingomonas sp.',\n",
        "    'uncultured bacterium': 'Other bacterium',\n",
        "    'bacterium': 'Other bacterium'\n",
        "}\n",
        "nr_metagenomes_by_organism_cleaned['Organism'] = nr_metagenomes_by_organism_cleaned['Organism'].replace(map)\n",
        "nr_metagenomes_by_organism_cleaned = nr_metagenomes_by_organism_cleaned.groupby('Organism')['Reference'].nunique()\n",
        "\n",
        "nr_metagenomes_by_organism_cleaned_normalized = nr_metagenomes_by_organism_cleaned[nr_metagenomes_by_organism_cleaned.values > 4]\n",
        "\n",
        "nr_metagenomes_by_organism_cleaned_normalized_plot = plt.bar(nr_metagenomes_by_organism_cleaned_normalized.index, nr_metagenomes_by_organism_cleaned_normalized.values, color='orange')\n",
        "plt.xticks(rotation=90, fontsize=6)\n",
        "plt.xlabel(\"Organismo\")\n",
        "plt.ylabel(\"Cantidad de metagenomas\")\n",
        "plt.title(\"Cantidad de metagenomas por organismo\")\n",
        "plt.savefig(\"metagenomes_by_organism_cleaned_normalized.png\", dpi=300, bbox_inches='tight')\n",
        "plt.show(nr_metagenomes_by_organism_cleaned_normalized_plot)\n",
        "display(nr_metagenomes_by_organism_cleaned_normalized)\n",
        "\n",
        "#without normalization\n",
        "nr_metagenomes_by_organism_cleaned_plot = plt.bar(nr_metagenomes_by_organism_cleaned.index, nr_metagenomes_by_organism_cleaned.values, color='orange')\n",
        "plt.xticks(rotation=90, fontsize=6)\n",
        "plt.xlabel(\"Organismo\")\n",
        "plt.ylabel(\"Cantidad de metagenomas\")\n",
        "plt.title(\"Cantidad de metagenomas por organismo\")\n",
        "plt.savefig(\"metagenomes_by_organism_cleaned.png\", dpi=300, bbox_inches='tight')\n",
        "plt.show(nr_metagenomes_by_organism_cleaned_plot)\n",
        "\n",
        "#reverse plot\n",
        "\n",
        "nr_metagenomes_by_organism_cleaned_reversed = nr_metagenomes_by_organism_cleaned.groupby('Reference')['Organism'].nunique()\n",
        "\n",
        "nr_metagenomes_by_organism_cleaned_reversed = nr_metagenomes_by_organism_cleaned_reversed[nr_metagenomes_by_organism_cleaned_reversed.values > 3]\n",
        "nr_metagenomes_by_organism_cleaned_reversed = nr_metagenomes_by_organism_cleaned_reversed[nr_metagenomes_by_organism_cleaned_reversed.values < 200]\n",
        "\n",
        "\n",
        "nr_metagenomes_by_organism_cleaned_reversed_plot = plt.bar(nr_metagenomes_by_organism_cleaned_reversed.index, nr_metagenomes_by_organism_cleaned_reversed.values, color='orange')\n",
        "plt.xticks(rotation=90, fontsize=6)\n",
        "plt.xlabel(\"Metagenome\")\n",
        "plt.ylabel(\"Number of organisms\")\n",
        "plt.title(\"Number of organisms by metagenome\")\n",
        "plt.savefig(\"metagenomes_by_organism_cleaned_reversed.png\", dpi=300, bbox_inches='tight')\n",
        "plt.show(nr_metagenomes_by_organism_cleaned_reversed_plot)\n",
        "display(nr_metagenomes_by_organism_cleaned_reversed)"
      ],
      "metadata": {
        "id": "bkJMbPd6KkWX"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#list of metagenomes comparison\n",
        "\n",
        "metagenomes_list_comparison = pd.merge(nr_pathways_in_metagenome_reverse, nr_metagenomes_by_organism_cleaned_reversed, on='Reference', how='inner')\n",
        "display(metagenomes_list_comparison)"
      ],
      "metadata": {
        "id": "gKlA7if-KkRL"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#plot Metagenome distribution by environment from TSV file from PROMOTOR\n",
        "\n",
        "df_environments = pd.read_csv(\"metagenomes.tsv\", sep=\"\\t\")\n",
        "\n",
        "nr_metagenomes_by_environments = df_environments.groupby('category')['environment'].nunique()\n",
        "\n",
        "display(nr_metagenomes_by_environments)\n",
        "\n",
        "nr_metagenomes_by_environments_plot = plt.bar(nr_metagenomes_by_environments.index, nr_metagenomes_by_environments.values, color='orange')\n",
        "plt.xticks(rotation=90, fontsize=6)\n",
        "plt.xlabel(\"Medio ambiente\")\n",
        "plt.ylabel(\"Cantidad de metagenomas\")\n",
        "plt.title(\"Cantidad de metagenomas por medio ambiente\")\n",
        "plt.savefig(\"metagenomes_by_environment.png\", dpi=300, bbox_inches='tight')\n",
        "plt.show(nr_metagenomes_by_environments_plot)"
      ],
      "metadata": {
        "id": "MXouEJzsKkCe"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#plots for species analysis in mgr\n",
        "#PLOT genes_by_organism\n",
        "\n",
        "#cleaned by genera version\n",
        "\n",
        "nr_genes_in_organism_cleaned = df_dict['organism_pathway_df_combined']\n",
        "\n",
        "map = {\n",
        "    'Acinetobacter baumannii': 'Acinetobacter sp.',\n",
        "    'Acinetobacter calcoaceticus': 'Acinetobacter sp.',\n",
        "    'Alcaligenes faecalis': 'Alcaligenes sp.',\n",
        "    'Bacillus subtilis': 'Bacillus sp.',\n",
        "    'Bacillus licheniformis': 'Bacillus sp.',\n",
        "    'Bacillus amyloliquefaciens': 'Bacillus sp.',\n",
        "    'Bacillus velezensis': 'Bacillus sp.',\n",
        "    'Burkholderia cepacia': 'Burkholderia sp.',\n",
        "    'Comamonas testosteroni': 'Comamonas sp.',\n",
        "    'Escherichia coli': 'Escherichia sp.',\n",
        "    'Methylobacterium extorquens': 'Methylobacterium sp.',\n",
        "    'Methanosarcina barkeri': 'Methanosarcina sp.',\n",
        "    'Methanosarcina mazei': 'Methanosarcina sp.',\n",
        "    'Methanosarcina siciliae': 'Methanosarcina sp.',\n",
        "    'Methanosarcina thermophila': 'Methanosarcina sp.',\n",
        "    'Methylosinus sporium': 'Methylosinus sp.',\n",
        "    'Methylosinus trichosporium': 'Methylosinus sp.',\n",
        "    'Pseudomonas aeruginosa': 'Pseudomonas sp.',\n",
        "    'Pseudomonas fluorescens': 'Pseudomonas sp.',\n",
        "    'Pseudomonas pseudoalcaligenes': 'Pseudomonas sp.',\n",
        "    'Pseudomonas putida': 'Pseudomonas sp.',\n",
        "    'Pseudomonas stutzeri': 'Pseudomonas sp.',\n",
        "    'Pseudomonas mendocina': 'Pseudomonas sp.',\n",
        "    'Pseudomonas viridiflava': 'Pseudomonas sp.',\n",
        "    'Ralstonia eutropha': 'Ralstonia sp.',\n",
        "    'Ralstonia pickettii': 'Ralstonia sp.',\n",
        "    'Rhodococcus erythropolis': 'Rhodococcus sp.',\n",
        "    'Rhodococcus jostii': 'Rhodococcus sp.',\n",
        "    'Rhodococcus opacus': 'Rhodococcus sp.',\n",
        "    'Rhodococcus rhodochrous': 'Rhodococcus sp.',\n",
        "    'Sphingomonas paucimobilis': 'Sphingomonas sp.',\n",
        "    'Salmonella enterica subsp. enterica serovar Enteritidis': 'Salmonella sp.',\n",
        "    'Salmonella enterica subsp. enterica serovar Montevideo str.': 'Salmonella sp.',\n",
        "    'uncultured bacterium': 'Other bacterium',\n",
        "    'bacterium': 'Other bacterium'\n",
        "}\n",
        "nr_genes_in_organism_cleaned['Organism'] = nr_genes_in_organism_cleaned['Organism'].replace(map)\n",
        "nr_genes_in_organism_cleaned = nr_genes_in_organism_cleaned.groupby('Organism')['Gene_GI'].nunique()\n",
        "\n",
        "nr_genes_in_organism_cleaned_normalized = nr_genes_in_organism_cleaned[nr_genes_in_organism_cleaned.values > 10]\n",
        "\n",
        "nr_genes_in_organism_cleaned_normalized_plot = plt.bar(nr_genes_in_organism_cleaned_normalized.index, nr_genes_in_organism_cleaned_normalized.values, color='orange')\n",
        "plt.xticks(rotation=90, fontsize=6)\n",
        "plt.xlabel(\"Organismo\")\n",
        "plt.ylabel(\"Cantidad de genes\")\n",
        "plt.title(\"Cantidad de genes por organismo\")\n",
        "plt.savefig(\"genes_by_organism_cleaned_normalized.png\", dpi=300, bbox_inches='tight')\n",
        "plt.show(nr_genes_in_organism_cleaned_normalized_plot)\n",
        "display(nr_genes_in_organism_cleaned_normalized)"
      ],
      "metadata": {
        "id": "x88koR-PKjw-"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#df_environments\n",
        "df_ncbi_choose = pd.read_csv(\"unikalne_wartosci.tsv\", sep=\"\\t\")\n",
        "df_ncbi_choose.columns = ['reference']\n",
        "\n",
        "ncbi_environment_combined = pd.merge(df_ncbi_choose, df_environments, on='reference', how='inner')\n",
        "ncbi_environment_combined = ncbi_environment_combined[['reference', 'category']]\n",
        "display(ncbi_environment_combined)\n",
        "count_ncbi_environment = ncbi_environment_combined.groupby('category')['reference'].nunique()\n",
        "display(count_ncbi_environment)"
      ],
      "metadata": {
        "id": "uoVsSc4KdogT"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}